bindings/c/tools/gen_header.py

```diff
@@ -12,98 +15,106 @@
-		in_block = True
-		arg_type = 'int (0 or 1)'
-	elif l == 'enum int_types':
-		in_block = True
-		arg_type = 'int'
-	elif not in_block:
-		continue
-	else:
-
-		if l.startswith('#if'):
-			in_define_block = True
-			continue
-
-		setting = l.replace(',','').split('=')[0].strip()
-
-		if setting.endswith('_internal'): continue
-
-		cpp.write('\t\tcase SET_%s: return sp::%s;\n' % (setting.upper(), setting))
-		if first:
-			header.write('	SET_%s = 0x200, // %s\n' % (setting.upper(), arg_type))
-			first = False
-		else:
-			header.write('	SET_%s, // %s\n' % (setting.upper(), arg_type))
-
-cpp.write('''		default:
-			// ignore unknown tags
-			return -1;
-	}
+for line in f:
+    line = line.strip()
+
+    if in_define_block:
+        if line.startswith("#endif"):
+            in_define_block = False
+        continue
+
+    if line == "{" or line == "" or line.startswith("//"):
+        continue
+
+    if in_block and line == "};":
+        in_block = False
+    elif line == "enum string_types":
+        in_block = True
+        arg_type = "char const*"
+    elif line == "enum bool_types":
+        in_block = True
+        arg_type = "int (0 or 1)"
+    elif line == "enum int_types":
+        in_block = True
+        arg_type = "int"
+    elif not in_block:
+        continue
+    else:
+
+        if line.startswith("#if"):
+            in_define_block = True
+            continue
+
+        setting = line.replace(",", "").split("=")[0].strip()
+
+        if setting.endswith("_internal"):
+            continue
+
+        cpp.write("\t\tcase SET_%s: return sp::%s;\n" % (setting.upper(), setting))
+        if first:
+            header.write("\tSET_%s = 0x200, // %s\n" % (setting.upper(), arg_type))
+            first = False
+        else:
+            header.write("\tSET_%s, // %s\n" % (setting.upper(), arg_type))
+
+cpp.write(
+    """\t\tdefault:
+\t\t\t// ignore unknown tags
+\t\t\treturn -1;
+\t}
 }
 
 lt::settings_pack make_settings(int tag, va_list lp)
 {
-	lt::settings_pack ret;
-	using sp = lt::settings_pack;
-	while (tag != 0)
-	{
-		int const key = settings_key(tag);
-		switch (key & lt::settings_pack::type_mask)
-		{
-			case sp::string_type_base:
-				ret.set_str(key, va_arg(lp, char*));
-				break;
-			case sp::int_type_base:
-				ret.set_int(key, va_arg(lp, int));
-				break;
-			case sp::bool_type_base:
-				ret.set_bool(key, va_arg(lp, int));
-				break;
-		}
-		tag = va_arg(lp, int);
-	}
-	return ret;
+\tlt::settings_pack ret;
+\tusing sp = lt::settings_pack;
+\twhile (tag != 0)
+\t{
+\t\tint const key = settings_key(tag);
+\t\tswitch (key & lt::settings_pack::type_mask)
+\t\t{
+\t\t\tcase sp::string_type_base:
+\t\t\t\tret.set_str(key, va_arg(lp, char*));
+\t\t\t\tbreak;
+\t\t\tcase sp::int_type_base:
+\t\t\t\tret.set_int(key, va_arg(lp, int));
+\t\t\t\tbreak;
+\t\t\tcase sp::bool_type_base:
+\t\t\t\tret.set_bool(key, va_arg(lp, int));
+\t\t\t\tbreak;
+\t\t}
+\t\ttag = va_arg(lp, int);
+\t}
+\treturn ret;
```

Comment:
what happened here?

Reply:
As I mentioned in the description, tools don't like hard tabs in python files very much.

Here, I believe `black` converted all the hard tabs into spaces in the indented python code. Then, `flake8` interpreted the hard tabs in the multiline string as being "indentation", and complained you shouldn't mix hard tabs and spaces for indentation.

It's also burdensome to figure out how to configure an editor to work with hard tabs in just this one file.

`\t` works fine as a tab character for multiline strings, so I replaced the hard tabs by hand.

Other options:
 * We could leave this file on the exclusion lists, so `black` won't autoformat it (I'm not sure `flake8` can easily be configured to ignore mixed-space warnings in this specific file, but I'll see what I can do)
 * We could turn the hard tabs in generated files into spaces

Reply:
so, github prints tab characters as "\t"? All C++ code use tabs, I don't think they are printed as "\t", that's why I thought these were literally backslash-t.

Reply:
You can also replace tabs with `%s/\t/    /g`

Reply:
They *are* literally backslash-t

Reply:
But the backslash-t is in a multi-line quoted string

Reply:
ah, I see. flake8 has opinions about indentation in strings too :)

Reply:
Right. See PyCQA/pycodestyle#45 for some more info on it

Reply:
that thread suggests putting `# noqa` after the string would also fix it. I would prefer that if it works

Reply:
Ah, just saw this comment. I applied this change, narrowly filtering just the whitespace complaints. What do you think?